napari/components/add_layers_mixin.py

```diff
@@ -436,6 +438,12 @@ def add_labels(
             Labels data as an array or multiscale.
         num_colors : int
             Number of unique colors to use in colormap.
+        properties : dict {str: array (N,)}, DataFrame
+            Properties for each label. Each property should be an array of length
+            N, where N is the number of labels.
+        label_index : dict {int: int}
+            Dictionary mapping labels (arbitrary integers) to row indices
+            (sequential integers in 0..N not inclusive).
```

Comment:
Add a disclaimer here like: "A label value not found in the index keys will be considered as having no properties associated with it."

Reply:
Does every row need to be mapped to? If so then maybe an idea is to have `label_index` be a length `N` list where each value corresponds to a label. We could even construct this dict from that list if we wanted faster indexing/ convenience. So you'd have

```python
[1, 10, 100] # equivalent to
{'1': 0, '10':1, '100':2}
```

Feel free to disregard

Reply:
Going further, could you actually then just have `label_index` as a special column within `properties`? Basically each row of properties would tell us which label it wanted to correspond to, and if no special column provided we just do default mapping.

Reply:
@sofroniewn well, the thing is we need the dictionary for fast lookup. So we'll have to create the attribute anyway. However, your suggestion actually makes a lot of sense in terms of reducing the API. My preference would be for the column to be called `index`, as that is the default in pandas if you do `reset_index()` with an unnamed index.

We would also need to update `dataframe_to_properties` to include this convention: if there is no column named index *and* the index has type "int", use the pandas index.

Reply:
Yeah, I think we'd pay the price of creating that (private) attribute, but for the gain of less API. I like a convention using the word `index`, and think that's a reasonable update to `dataframe_to_properties`

We'd then have to decide for our defaults would the first row map to background? With this API I think not. We could have it map to `1` by default. But again, open to ideas/ suggestions

Reply:
Let's go with 0 default. Corresponds to NumPy indexing.

Reply:
OK so just double checking the upshot from this conversation:

- Make `label_index` private, rename it to `index`
- When being passed a `properties` dict, check if it has an `index` column
- If it does, pull it into our `index` dict
- If it doesn't, use the pandas index column

> We would also need to update dataframe_to_properties to include this convention: if there is no column named index and the index has type "int", use the pandas index.

@jni I'm a bit confused by this. if there is no column named index how are we checking it's type int? Did you mean to say if there IS an index and it's 0..N, we just use pandas index?